man/sycersion.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SYCERSION" "1" "January 2025" "" ""
.
.SH "NAME"
\fBsycersion\fR \- manage application versions according to semantic versioning
.
.SH "SYNOPSIS"
.
.nf

`sycersion` `\-\-init`
`sycersion` `\-i`|`\-\-info`
`sycersion` `\-\-inc major|minor|patch`
`sycersion` `\-\-set major\.minor\.patch[+build|\-pre\-release+build]`
`sycersion` `\-p`|`\-\-pre` `pre\-release`
`sycersion` `\-b`|`\-\-build` `build`
`sycersion` `\-c`|`\-\-compare` `version`
`sycersion` `\-h`|`\-\-help`
.
.fi
.
.SH "Description"
\fBsycersion\fR is a command\-line tool to manage application versions following the guiding principles of \fIsemantic versioning\fR found at semver\.org \fIhttps://semver\.org\fR\.
.
.P
\fBsycersion\fR has a default start version of 0\.1\.0\. This can be activitated with the initalization running \fBsycersion \-\-init\fR\. During this intialization another version can be chosen\. The version is saved to the \fIversion\-file\fR and also the \fIconfiguration\-directory\fR is created\. While initialization \fBsycersion\fR additionally searches for files that contain a version according to semantic versioning and the user can select out of the list of files which version out of that files to intially start with\.
.
.P
The \fIversion\-file\fR and \fIconfiguration\-directory\fR are also created it the user is setting an initial version with \fBsycersion \-\-set 0\.1\.1\fR\.
.
.P
If the version is updated the version is written into the \fIversion\-file\fR\. Within the application the version can be read from the file and displayed somewhere in the application\'s UI\.
.
.P
\fBsycersion\fR evaluates version numbers provided that they comply to semantic versioning\.
.
.P
If \fBsycersion\fR is used within a \fIRuby\fR application, it can be added to the development section within the Gemfile\.
.
.SH "FILES"
.
.nf

`\.sycersion` is the configuration directory
`\.sycersion/environment\.yml` holds the version file and the version
`\.sycersion/version` contains the version
.
.fi
.
.SH "OPTIONS"
.
.IP "\(bu" 4
\fB\-\-init\fR Initializes the \fBsycersion\fR environment with \fIversion\fR and \fIversion\-file\fR
.
.IP "\(bu" 4
\fB\-i\fR, \fB\-\-info VERSION|SUMMARY\fR Shows the current \fIversion\fR (default) or additionally information about the \fIversion\-file\fR, the \fIconfiguration\-directory\fR and \fIcode\fR how to read the version within an application\.
.
.IP "\(bu" 4
\fB\-\-inc MAJOR|MINOR|PATCH\fR increments the major, minor, patch part of the version\.
.
.IP "\(bu" 4
\fB\-\-set MAJOR\.MINOR\.PATCH[+BUILD|\-PRE_RELEASE[+BUILD]]\fR sets the version where all three version parts need to be provided, major\.minor\.patch with optionally adding a pre\-release part and/or a build part\.
.
.IP "\(bu" 4
\fB\-\-pre PRE\-RELEASE\fR Set the pre\-release part in the version
.
.IP "\(bu" 4
\fB\-\-build BUILD\fR Set the build part in the version
.
.IP "\(bu" 4
\fB\-c\fR, \fB\-\-compare VERSION\fR Compares the current version with the version provided following the semver precedence guidline
.
.IP "\(bu" 4
\fB\-\-help\fR Prints the command line help
.
.IP "" 0
.
.SH "EXAMPLES"
Initialize \fBsycersion\fR
.
.IP "" 4
.
.nf

$ sycersion \-\-init
.
.fi
.
.IP "" 0
.
.P
Set the version with pre\-release
.
.IP "" 4
.
.nf

$ sycersion \-\-set 0\.1\.1\-beta\.1\.0
.
.fi
.
.IP "" 0
.
.P
Set the version with a build
.
.IP "" 4
.
.nf

$ sycersion \-\-set 0\.1\.1+build\-a\.1
.
.fi
.
.IP "" 0
.
.P
Set the version with a pre\-release and a build
.
.IP "" 4
.
.nf

$ sycersion \-\-set 0\.1\.1\-beta\.1\.0+build\-a\.1
.
.fi
.
.IP "" 0
.
.P
Show the current version
.
.IP "" 4
.
.nf

$ sycersion \-i
0\.1\.1\-beta\.1\.0+build\-a\.1
.
.fi
.
.IP "" 0
.
.P
Increment minor part of the version 0\.1\.1, which will lead to 0\.2\.0
.
.IP "" 4
.
.nf

$ sycersion \-\-inc minor
$ sycersion \-i
0\.2\.0
.
.fi
.
.IP "" 0
.
.P
When increasing one of major, minor or patch, the subsequent numbers are eather reset to 0 or pre\-release and build are removed\.
.
.P
Compare the assumed current version 0\.1\.1 to the provided version
.
.IP "" 4
.
.nf

$ sycersion \-c 0\.1\.1\-alpha
0\.1\.1 > 0\.1\.1\-alpha
.
.fi
.
.IP "" 0
.
.SH "ENVIRONMENT"
.
.IP "\(bu" 4
SYC_DEBUG=true Will display erros and show contents of the parsed options\. To be called as \fBSYC_DEBUG=true sycersion [options]\fR
.
.IP "" 0



lib/sycersion.rb
# frozen_string_literal: true

require_relative 'sycersion/version'
require_relative 'sycersion/options'
require_relative 'sycersion/version_environment'
require_relative 'sycersion/version_setter'
require_relative 'sycersion/version_incrementer'
require_relative 'sycersion/version_info'
require_relative 'sycersion/version_compare'
require_relative 'sycersion/semver'

module Sycersion
  class Error < StandardError; end

  # Runs the program based on the arguments provided at the commandline
  class Runner
    def initialize(argv)
      @options = Options.new(argv).options
      pp @options if ENV['SYC_DEBUG']
    end

    def run
      environment_settings
      version_manipulation
      version_compare
      configuration_information
    end

    def environment_settings
      Sycersion::VersionEnvironment.new.setup if @options[:init]
    end

    def version_manipulation
      case @options.keys
      when [:set]
        Sycersion::VersionSetter.new.version = (@options[:set])
      when [:pre_release]
        Sycersion::VersionSetter.new.pre_release = (@options[:pre_release])
      when [:build]
        Sycersion::VersionSetter.new.build = (@options[:build])
      when [:inc]
        Sycersion::VersionIncrementer.new.increment(@options[:inc])
      end
    end

    def version_compare
      return unless @options[:compare]

      @options[:compare].shift
      puts Sycersion::VersionCompare.new.compare(@options[:compare])
    end

    def configuration_information
      puts Sycersion::VersionInfo.new.process(@options) if @options[:info]
    end
  end
end
